### DISCUSSION TITLE
Server-side fetch returns a 403 on build

### DISCUSSION THREAD

**Discussion by conorhub** [vercel/next.js#45177] [Help] [answered]
I'm migrating an app from **create-react-app** to **nextjs**. When fetching an item from the db backend I used the following in the original **create-react-app**:

    export async function fetchSingleItem(item_id) {
    
    	try {
    
    		let fields = [
    			'item_title',
    			'item_description',
    		]
    
    		let url = api_server_url + "/items/" + item_id + "?fields=" + fields;
    		const fetchit = await fetch(url);
    		if (!fetchit.ok) {
    			throw "Sorry, failed to fetch item data, there was some problem";
    		}
    		const fetchitjson = fetchit.json();
    		return fetchitjson;
    
    	}
    	catch(error) {
    
    		..error handling ...
    		
    	}
    
    }

This still works as intended in nextjs as a server-side component when running in *dev* but not in *build*.

When I attempt to *build*, it returns a 403.

Looking at the [docs][1] (I'm using nextjs13 with experimental 'app' directory) it suggests that using fetch for server-side is fine. I'm not using typescript.

### Backend

The backend is set up to allow cors and all origins. I still assume this is working as intended as, in the create-react-app, I had no problem running this same function from localhost and on a production server with a different domain.

### Question

I've no idea what's causing fetch to 403 when building, but not when running in *dev*. I've tried various fetch options and none worked. I don't need any additional options when running in *dev* for fetch to pull date as intended, so I'm not sure if that's the problem with *build*.

  [1]: https://beta.nextjs.org/docs/data-fetching/fetching

**Comment by icyJoseph** [comment, marked as answer]
Could it be that a user agent or something like that is the cause of the issue? Are you able to debug from your server? I am not sure if the fetch during dev mode includes some data, whereas the production env fetch doesn't?

**Reply by conorhub** [reply]
> Could it be that a user agent or something like that is the cause of the issue? 

I don't believe so, unless it's something nextjs changes by default? I haven't messed with any configs in this area. Fetch is throwing an error when building both locally and deploying to production server.

> Are you able to debug from your server?

Not in any great depth. I'm using a headless cms running on a nodejs stack. It has a basic log and, when I attempt to build, it shows it receives a GET request from the fetch I'm having issues with. But I don't believe it's possible to see logs further than that.

I'm still presuming fetch being run during production build is doing something different than when in development. Almost all answers relate to the former data-fetching functions of nextjs, pre v13. I'm trying to see if there's something relevant in a similar scenario in those but can't find it yet.

I'm hoping there's a header I can change in fetch() that will solve all of this but haven't found the right combination yet as I'm only guessing, to be honest.

**Reply by conorhub** [reply]
> I've no idea what's causing fetch to 403 when building, but not when running in _dev_...

I've discovered a cause but it's led to another problem I don't understand...

It's not fetch that's failing on build. It's params. Params are being read correctly when running as dev, as they are being fed specific routing info such as site/[firstparam]/[secondparam]

But are not working on build. I'm unsure how dynamic routing is handled on build when a user hasn't selected a specific route yet. I beleived that either it would have a value given to it (as it does in dev mode) or be undefined. Yet, when it's building, if I log params, it says they have values and that they're escaped versions of the key name, such as:

params.secondparam = '%5Bsecondparam%5D'

Which I don't understand?

This was getting missed when I was checking that params weren't undefined before passing them to functions.

**Reply by icyJoseph** [reply]
Aha, but when you build dynamic routes, like in pages directory, you gotta tell Next.js that either: 

- all paths are known at build time
- all paths will be defined at run time
- some paths are known at build time, but others will be defined at runtime 

Which one do you need?

**Reply by conorhub** [reply]
The rational of what you're suggesting makes sense. I'm just surprised it's not mentioned as a scenario in the documentation for params (or that I missed it). Thanks for the assistance, it's appreciated

Params, in my case, are basically ids for database items in the backend along with an seo friendly slug.

They change and grow long after the build, so they will only be known when running.
